packages/thirdweb/src/react/web/ui/prebuilt/thirdweb/BuyDirectListingButton/index.tsx

```diff
@@ -0,0 +1,197 @@
+import { useSendAndConfirmTransaction } from "../../../../../core/hooks/transaction/useSendAndConfirmTransaction.js";
+import { useActiveAccount } from "../../../../../core/hooks/wallets/useActiveAccount.js";
+import { TransactionButton } from "../../../TransactionButton/index.js";
+
+export type BuyDirectListingButtonProps = Omit<
+  TransactionButtonProps,
+  "transaction"
+> & {
+  /**
+   * The contract address of the Marketplace v3 contract.
+   */
+  contractAddress: string;
+
+  /**
+   * The chain which the Drop contract is deployed on
+   */
+  chain: Chain;
+
+  /**
+   * thirdweb Client
+   */
+  client: ThirdwebClient;
+
+  /**
+   * ID of the marketplace's DirectListing
+   */
+  listingId: bigint;
+
+  /**
+   * Qty to buy (optional)
+   *
+   * - For ERC721 listing: the `quantity` is always hard-coded to 1n - passing this props doesn't do anything
+   *
+   * - For ERC1155 listing: the `quantity` defaults to the quantity of the listing if not specified.
+   *
+   * The component will also throw an error if  you pass a `quantity` and it's greater than the listing's quantity
+   */
+  quantity?: bigint;
+};
+
+/**
+ * This button is used with thirdweb Marketplace v3 contract, for buying NFT(s) from a listing.
+ *
+ * Under the hood, it prepares a transaction using the [`buyFromListing` extension](https://portal.thirdweb.com/references/typescript/v5/marketplace/buyFromListing)
+ * and then pass it to a <TransactionButton />
+ *
+ * Since it uses the TransactionButton, it can take in any props that can be passed
+ * to the [`TransactionButton`](https://portal.thirdweb.com/references/typescript/v5/TransactionButton)
+ *
+ *
+ * @param props
+ * @returns <TransactionButton />
+ *
+ * @example
+ * ```tsx
+ * import { BuyDirectListingButton } from "thirdweb/react";
+ *
+ * <BuyDirectListingButton
+ *   contractAddress="0x..." // contract address of the marketplace v3
+ *   chain={...} // the chain which the marketplace contract is deployed on
+ *   client={...} // thirdweb client
+ *   listingId={100n} // the listingId or the item you want to buy
+ *   quantity={1n} // optional - see the docs to learn more
+ * >
+ *   Buy NFT
+ * </BuyDirectListingButton>
+ * ```
+ */
+export function BuyDirectListingButton(props: BuyDirectListingButtonProps) {
+  const {
+    contractAddress,
+    listingId,
+    children,
+    chain,
+    client,
+    quantity,
+    payModal,
+  } = props;
+  const defaultPayModalMetadata = payModal ? payModal.metadata : undefined;
+  const account = useActiveAccount();
+  const contract = getContract({
+    address: contractAddress,
+    client,
+    chain,
+  });
+
+  const { data: payMetadata } = useReadContract(getPayMetadata, {
+    contract,
+    listingId,
+    queryOptions: {
+      enabled: !defaultPayModalMetadata,
+    },
+  });
+
+  const { mutateAsync } = useSendAndConfirmTransaction();
+
+  async function prepareBuyTransaction() {
```

Comment:
could use a `useCallback` for this so it doesnt recreate on every render, but only when props change

Reply:
Done!